Add query results cache support for cardinality analysis API

pracucci opened the pull request:
#### What this PR does
We're seeing some customers issuing the same cardinality analysis query over and over, few seconds apart from each other (likely coming from a script or internal tooling, given the user agent). These requests can be pretty heavy for the ingester, in terms of CPU utilization and memory allocations.

The cardinality analysis API, whose returned value is already a close-to-reality approximation today (but not a 100% accurate count), doesn't need to guarantee read-after-write and some short cache TTL could be applied to it. After discussing it with few people at Grafana Labs, in this PR I'm introducing the optional support to cache cardinality query results (disabled by default, TTL configurable on a per-tenant basis).

Note to reviewers:
- dskit has been updated to get a couple of more methods to `cache.InstrumentedCacheMock` ([PR](https://github.com/grafana/dskit/pull/316))
- I manually tested it and looks doing its job

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pstibrany commented on pkg/frontend/querymiddleware/cardinality_query_cache.go:
```diff
@@ -0,0 +1,186 @@
+	}
+
+	// Skip the cache if disabled for the tenant.
+	cacheTTL := validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, c.limits.ResultsCacheTTLForCardinalityQuery)
```
nit: If tenantIDs has two tenants and one of them has caching disabled, but other one doesn't, this will return cacheTTL from tenant that doesn't have it disabled. That seems like a bug to me.

pracucci replied:
Good point. It's not a nit, but a bug! Will add a test to cover it too.

pracucci replied:
Test added in https://github.com/grafana/mimir/pull/5212/commits/768a4a5a508add98e1d6d69589ef82f80cc0bd00.

pstibrany commented on pkg/frontend/querymiddleware/cardinality_query_cache.go:
```diff
@@ -0,0 +1,186 @@
+		level.Info(spanLog).Log("msg", "skipped cardinality query caching because failed to parse the request", "err", err)
+
+		// We skip the caching but let the downstream try to handle it anyway,
+		// since it's not our responsibility to decide how to response in this case.
```
```suggestion
		// since it's not our responsibility to decide how to respond in this case.
```

pstibrany commented on pkg/frontend/querymiddleware/cardinality_query_cache.go:
```diff
@@ -0,0 +1,186 @@
+	return
+}
+
+func isCardinalityQueryResponseCachable(res *http.Response) bool {
```
suggestion: `isCardinalityQueryResponseCachable` => `isCardinalityQueryResponseCacheable`

pstibrany commented on pkg/frontend/querymiddleware/model.proto:
```diff
@@ -116,3 +116,25 @@ message Hints {
+message CachedHTTPResponse  {
+  // CacheKey contains the non-hashed cache key, used to guarantee there haven't
+  // been any hashed cache key collision.
+  string CacheKey = 1;
```
nit: Fields in protobuf should be lowercase. Although some (not all) messages in this file don't do that, let's not repeat that mistake.

Applies to `CachedHTTPHeader` too.

pracucci replied:
👍 TIL

pstibrany replied:
https://protobuf.dev/programming-guides/style/#message-field-names :)

pracucci replied:
Aha! It's lower_snake_case for fields. Fixed in https://github.com/grafana/mimir/pull/5212/commits/6e436ed5c18499c2071c738a770f0c3f80ba043c

pstibrany commented on pkg/frontend/querymiddleware/model.proto:
```diff
@@ -116,3 +116,25 @@ message Hints {
+  bytes Body = 3;
+
+  // Headers contains the HTTP response headers.
+  repeated CachedHTTPHeader Headers = 4;
```
nit: not too important, but logical order of fields would be: status code, headers, body, same as in HTTP request.

pracucci replied:
Ok :)

pstibrany commented on pkg/frontend/querymiddleware/model_extra.go:
```diff
@@ -419,3 +422,47 @@ func (resp *PrometheusResponse) minTime() int64 {
+	// otherwise the caller will get a response with a closed Body.
+	res.Body = io.NopCloser(bytes.NewBuffer(body))
+
+	// When preallocating the slice we assume that header as 1 value (which is the most common case).
```
This comment seems obsolete.

Food for thought: we're caching HTTP responses generated by Mimir components. Afaik Mimir doesn't set any HTTP headers with multiple values. So we could simplify our code and just use `map<string, string> headers` in protobuf definition. Although not strictly correct for general case, good-enough for Mimir use.

(I already know the answer, but worth pointing out)

pracucci replied:
> This comment seems obsolete.
Why obsolete? 

>So we could simplify our code and just use map<string, string> headers in protobuf definition.
>(I already know the answer, but worth pointing out)
Given it's doesn't extra complexity, I prefer to keep it consistent with the `http` package where you can have multiple values, given this model is a generic one.

pstibrany replied:
> Why obsolete?

I don't see any preallocation of slice for header values. 🤔 I'm not sure what that comment refers to.

pracucci replied:
The comment is about this pre-allocation:
```
headers := make([]*CachedHTTPHeader, 0, len(res.Header))
```

This pre-allocation is correct if we'll have 1 value for each header (common case).

pstibrany replied:
Ah, now I see what you mean. Assuming one value per header and N headers, we preallocate with N.

pstibrany commented on pkg/frontend/querymiddleware/cardinality_query_cache.go:
```diff
@@ -0,0 +1,186 @@
+
+	// Store the result in the cache.
+	if isCardinalityQueryResponseCachable(res) {
+		// The res.Body will be replaced after calling the following function.
```
I'd move this comment to `EncodeCachedHTTPResponse`, which should be documented as preserving `res.Body`.

pracucci replied:
Good point.

pstibrany commented on pkg/querier/cardinality/request.go:
```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package cardinality
```
Placing this under `pkg/querier/cardinality` seems strange due to "querier" part. Perhaps `pkg/cardinality` would be a better fit, so that `querymiddleware` doesn't "reach out" to querier packages.

pracucci replied:
Ok I can do it, but `querymiddleware` is already reaching `pkg/querier/...` subpackages.

pstibrany commented on pkg/querier/cardinality/request.go:
```diff
@@ -0,0 +1,225 @@
+	// Add label names.
+	for idx, name := range r.LabelNames {
+		if idx > 0 {
+			b.WriteRune(',')
```
When new characters are allowed in label names, it may be possible to construct conflicting requests (eg. request with label names `["a", "b"]` and request with single label `["a,b"]` would have same representation).

It may be safer to use `\0` as delimiter or some other value that will "for sure" never appear in label names or matchers.

pracucci replied:
Right. I've used `\0` and `\1` as param and values separator.

pstibrany commented on pkg/querier/cardinality/request.go:
```diff
@@ -0,0 +1,225 @@
+
+// DecodeLabelValuesRequest parses query params from GET requests and parses request body from POST requests.
+// TODO test (GET and POST)
+func DecodeLabelValuesRequest(r *http.Request) (*LabelValuesRequest, error) {
```
hint: if `DecodeLabelValuesRequest` works with `r.Form` only (type `url.Values`), you don't need to test both GET and POST here :)

Practically, testing with both GET/POST tests `r.ParseForm()` more than actual parameter parsing.

pracucci replied:
Well I just wanted to ensure we use `ParseForm()`. Unfortunately I read this comment too late, and I've already added that test (but it's a very small one).

pstibrany reviewed (approved):
Looks good, nice work!

pracucci commented:
Thanks @pstibrany for your review. I should have addressed all your comments and the TODO that I left when you did the 1st review. There are two things I would prefer to do in follow up PRs:

1. Improve HTTP API doc (since will require docs squad review, I prefer to not block this PR)
2. Try to improve metrics we have to compute the query results cache hit ratio (right now we can't distinguish between range and cardinality queries)

pstibrany reviewed (approved):
Looks good, thank you!
